groupbyattrsprocessor - part 2 - grouping logic (#1744)

pmm-sumo opened the pull request:
**Description:**

Introduces  processor for grouping the trace or log records by specified key values, and then extracting the attributes from the record to resource level.

This initiative is broken down into 3 separate PRs:

1. [Config](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1744) (already merged)

👉  2. The processor logic (this PR)

3. Enabling the component

_Example: consider having such input:_

```
InstrumentationLibraryLogs #0
LogRecord #0
Timestamp: 1606158664897894000
Severity:
ShortName:
Body: STRING("foo")
Attributes:
     -> action: STRING(login)
     -> namespace_name: STRING(kube-apiserver-docker-desktop)
     -> pod_name: STRING(kube-system)
```

With following config:

```
groupbyattrs:
  group_by_keys:
    - pod_name
    - namespace_name
```

The output is going to be:

```
Resource labels:
     -> namespace_name: STRING(kube-apiserver-docker-desktop)
     -> pod_name: STRING(kube-system)
InstrumentationLibraryLogs #0
LogRecord #0
Timestamp: 1606158617133076000
Severity:
ShortName:
Body: STRING("foo")
Attributes:
     -> action: STRING(login)
```

Since this is a first iteration of the processor and it generally applies for logs workflow with Kubernetes tagging (the attributes need to move from the record to the resource level) - the processor is introduced into opentelemetry-collector-contrib rather than the core repo.

**Link to tracking Issue:** [#1884] open-telemetry/opentelemetry-collector#1884

**Testing:** Unit tests added, manual tests performed

**Documentation:** Processor README.md included

pmm-sumo committed:
groupbyattrsprocessor - part 2 (grouping logic) #1744

pkositsyn commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// attributeGroups keeps all found grouping attributes, together with the matching records
+type attributeGroups struct {
+	entries []attributeEntry
+}
```
Is there a reason why this isn't `type attributeGroups []attributeEntry`?

pkositsyn commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -0,0 +1,160 @@
+		if matching {
+			v2, found := attrs2.Get(k1)
+			if found {
+				if !v1.Equal(v2) {
```
By the way, `Equal` now doesn't work with arrays (see implementation - always return false), but should by spec (and there must be no maps) - https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/common/common.md#attributes
Would you mind changing this as well or we can create an issue?

pmm-sumo replied:
I think that for the clarity a separate issue would be best

pkositsyn commented on processor/groupbyattrsprocessor/attribute_groups_test.go:
```diff
@@ -0,0 +1,68 @@
+	}
+)
+
+func BenchmarkAttrs(b *testing.B) {
+	lagAttrs.attributeEntry(groups[rand.Intn(count)], rs)
+}
```
Does Benchmark actually work like that without iterating until `b.N`?

pkositsyn commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -28,13 +28,111 @@ type groupbyattrsprocessor struct {
+		return groupedAnything, groupedAttrMap, attrMap
+	}
+
+	groupedAttrMap.Sort()
```
Is it really necessary? As I see from the doc it is needed only for testing

pkositsyn commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -28,13 +28,111 @@ type groupbyattrsprocessor struct {
+
+				groupedAnything, groupedAttrMap, nonGroupedAttrMap := gap.splitAttrMap(span.Attributes())
+				if groupedAnything {
+					mNumGroupedSpans.M(1)
```
If I'm not mistaken, these measurements are never recorded

pkositsyn commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -28,13 +28,111 @@ type groupbyattrsprocessor struct {
+			}
+		}
+
+		for _, ers := range extractedResourceSpans.entries {
+			ers.resourceSpans.CopyTo(rs)
+		}
```
CopyTo overwrites (at least should) the current value. So I don't see any point in ranging. Is it a mistake or not? Seems like this doesn't work with multiple ilss (not covered by tests). From my perspective, here should be appending of ilss to the zero resized resourceSpans.

pmm-sumo replied:
Yeah, a mistake! For some reason I thought CopyTo works like a ranged Upsert, but it just replaces the destination with copy of current values.

pkositsyn commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -28,13 +28,111 @@ type groupbyattrsprocessor struct {
+				groupedAnything, groupedAttrMap, nonGroupedAttrMap := gap.splitAttrMap(span.Attributes())
+				if groupedAnything {
+					mNumGroupedSpans.M(1)
+					overwriteAttributes(nonGroupedAttrMap, span.Attributes())
```
Does this processor really need to overwrite non-grouped attributes?

pmm-sumo replied:
Changed the flow to delete unnecessary attributes as suggested in another comment

pkositsyn commented on processor/groupbyattrsprocessor/processor_test.go:
```diff
@@ -0,0 +1,186 @@
+		ils.Spans().Append(span)
+	}
+
+	rs := pdata.NewResourceSpans()
+	rs.InstrumentationLibrarySpans().Append(ils)
+
+	traces := pdata.NewTraces()
+	traces.ResourceSpans().Append(rs)
```
Would be great to have traces with multiple rss and ilss

pkositsyn reviewed (commented):
While there is no review, I'd like to share my opinion on the PR

pkositsyn commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -0,0 +1,160 @@
+	resourceLogs  pdata.ResourceLogs
+}
+
+func instrumentationLibrariesEqual(il1, il2 pdata.InstrumentationLibrary) bool {
```
I'm not right - operator == compares the underlying pointers to 2 strings. In this case, what do you think about making`Equal` function on the struct instead of this?

pmm-sumo replied:
Do you mean adding `Equal` to the internal API of the models?

pkositsyn replied:
Yes, exactly. @bogdandrutu what do you think about that?

pkositsyn commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -0,0 +1,160 @@
+// given InstrumentatonLibrary. If nothing is found, it creates a new one
+func (ae *attributeEntry) instrumentationLibrarySpans(library pdata.InstrumentationLibrary) pdata.InstrumentationLibrarySpans {
+	ilss := ae.resourceSpans.InstrumentationLibrarySpans()
+	for i := 0; i < ilss.Len(); i++ {
+		ill := ilss.At(i)
+		if instrumentationLibrariesEqual(ill.InstrumentationLibrary(), library) {
+			return ill
+		}
+	}
```
Since operator == compares pointers - this isn't a working approach. Then either it can be applied to the struct with 2 fields  - name and version - which isn't good since the structure may be changed or actually leave it how it is done now

pmm-sumo commented:
> While there is no review, I'd like to share my opinion on the PR

I think reviewing the PR is always anticipated. Got some great suggestions, thank you @pkositsyn!

pkositsyn commented on processor/groupbyattrsprocessor/attribute_groups.go:
```diff
@@ -0,0 +1,160 @@
+		}
+	}
+
+	ills.Resize(ills.Len() + 1)
+	ill := ills.At(ills.Len() - 1)
+	library.CopyTo(ill.InstrumentationLibrary())
```
`Append` does not copy, just appends a new "reference" to the old object

pmm-sumo replied:
Each span can possibly have a different grouping attribute value, which might require copying InstrumentationLibrary for each one of them. I will update the flow and description to make it more clear what's going on here

tigrannajaryan commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -28,13 +28,111 @@ type groupbyattrsprocessor struct {
 	return td, nil
 }
 
 func (gap *groupbyattrsprocessor) ProcessLogs(_ context.Context, ld pdata.Logs) (pdata.Logs, error) {
```
The logic of what is happening in this function is not east to understand. Perhaps add comments or re-factor for more clarity.

tigrannajaryan commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -28,13 +28,111 @@ type groupbyattrsprocessor struct {
+				groupedAnything, groupedAttrMap, nonGroupedAttrMap := gap.splitAttrMap(log.Attributes())
+				if groupedAnything {
+					mNumGroupedLogs.M(1)
+					overwriteAttributes(nonGroupedAttrMap, log.Attributes())
```
Why do we overwrite non grouped attributes instead of deleting grouped attributes? Would it not have the same effect but be less work?

pmm-sumo replied:
I thought that would be less work to just copy the left attributes in bulk, but I was wrong. With deleting grouped attributes the code is actually simpler

bogdandrutu commented:
@pmm-sumo if stale bot closes this feel free to reopen the PR when comments are addressed.

pmm-sumo committed:
Address PR comments to groupbyattrsprocessor

pmm-sumo commented:
Many thanks @tigrannajaryan and @pkositsyn for the thorough and very helpful review. I addressed most of the comments, going to add more test cases and run several manual tests

pmm-sumo committed:
Refactor groupbyauthprocessor to cover more scenarios

pmm-sumo committed:
Add more attribute grouping tests

tigrannajaryan commented on processor/groupbyattrsprocessor/processor.go:
```diff
@@ -21,20 +21,117 @@ import (
+				groupedAnything, groupedAttrMap := gap.splitAttrMap(span.Attributes())
+				if groupedAnything {
+					mNumGroupedSpans.M(1)
+					// Some attributes are going to be moved from log record to resource level,
```
```suggestion
					// Some attributes are going to be moved from span to resource level,
```

pmm-sumo committed:
Update processor/groupbyattrsprocessor/processor.go

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>
